mne_bids/copyfiles.py

```diff
@@ -339,6 +341,102 @@ def copyfile_brainvision(vhdr_src, vhdr_dest, anonymize=None, verbose=False):
+        the dict: `daysback` is a required key, `keep_his` is an optional key.
+
+        `daysback` : int
+            Number of days by which to move back the recording date in time.
+            In studies with multiple subjects the relative recording date
+            differences between subjects can be kept by using the same number
+            of `daysback` for all subject anonymizations. `daysback` should be
+            great enough to shift the date prior to 1925 to conform with BIDS
+            anonymization rules. Due to limitations of the EDF/BDF format, the
+            year of the anonymized date will always be set to 1985 in the
+            'meas_date' region of the file. The correctly-shifted year will be
+            written to the 'recording info' region of the header, which may not
+            be parsed by all EDF/BDF reader software.
+
+        `keep_his` : bool
+            By default (False), all subject information next to the recording
+            date will be overwritten as well. If True, keep subject information
+            apart from the recording date. Participant names and birthdates
+            will always be anonymized if present, regardless of this setting.
+
+    See Also
+    --------
+    mne.io.anonymize_info
+
+    """
+    # Ensure source & destination extensions are the same
+    fname_src, ext_src = _parse_ext(src)
+    fname_dest, ext_dest = _parse_ext(dest)
+    if ext_src != ext_dest:
+        raise ValueError(f'Need to move data with same extension'
+                         f' but got "{ext_src}", "{ext_dest}"')
+
+    # Copy data prior to any anonymization
+    sh.copyfile(src, dest)
+
+    # Anonymize EDF/BDF data, if requested
+    if anonymize is not None:
+        if ext_src == '.bdf':
+            raw = read_raw_bdf(dest, preload=False, verbose=0)
+        else:
+            raw = read_raw_edf(dest, preload=False, verbose=0)
+
+        # Get patient info, recording info, and recording date
+        with open(dest, 'rb') as f:
+            f.seek(8)  # id_info field starts 8 bytes in
+            id_info = f.read(80).decode('ascii').rstrip()
+            rec_info = f.read(80).decode('ascii').rstrip()
+
+        # Parse metadata from file
+        if len(id_info) == 0 or len(id_info.split(' ')) != 4:
+            id_info = "X X X X"
+        if len(rec_info) == 0 or len(rec_info.split(' ')) != 5:
+            rec_info = "Startdate X X X X"
+        pid, sex, birthdate, name = id_info.split(' ')
+        startdate, admin_code, tech, equip = rec_info.split(' ')[1:5]
+
+        # Anonymize recording date, using 4-digit year from recording info
+        # if possible (regular EDF year only stores last two digits)
+        try:
+            real_year = datetime.datetime.strptime(startdate, "%d-%b-%Y").year
+            newdate = raw.info['meas_date'].replace(year=real_year)
+            raw.info['meas_date'] = newdate
+        except (ValueError, AttributeError):
+            pass
```

Comment:
`datetime.datetime.strptime(startdate, "%d-%b-%Y").year` looks like an error, it should be `datetime.strptime(startdate, "%d-%b-%Y").year` if you `import datetime from datetime`

Also, what exactly are you trying in this part of the code, and why could it fail with a ValueError ... and why with an AttributeError respectively?

Reply:
Good catch, was wondering why codecov was saying those lines weren't tested. As for the exceptions: ValueError is for if the startdate is an "X" or is otherwise improperly formatted (what datetime will throw if the provided string doesn't match the format string). AttributeError was in case raw.info['meas_date'] wasn't a valid datetime object, but in retrospect I can't think of a situation where it wouldn't be so I'll take it out.

Reply:
okay, please assert in the "except", that the raised exception is as you expected (wow, two similar words). --> and if it doesn't, raise an error instead of "passing".

Also, please add a comment on top of the "try:" so that future devs will know what happens.


finally --> is the meas_date *always* written with just two digits for year? Or can EDF+ or BDF write it with four digits?